L1Trigger/L1THGCalUtilities/plugins/CaloTruthCellsProducer.cc

```diff
@@ -0,0 +1,285 @@
+                                                  edm::EventSetup const&,
+                                                  HGCalTriggerGeometryBase const&) const;
+
+  typedef edm::AssociationMap<edm::OneToMany<CaloParticleCollection, l1t::HGCalTriggerCellBxCollection>> CaloToCellsMap;
+
+  bool makeCellsCollection_;
+  edm::EDGetTokenT<CaloParticleCollection> caloParticlesToken_;
+  edm::EDGetTokenT<l1t::HGCalTriggerCellBxCollection> triggerCellsToken_;
+  edm::EDGetTokenT<std::vector<PCaloHit>> simHitsTokenEE_;
+  edm::EDGetTokenT<std::vector<PCaloHit>> simHitsTokenHEfront_;
+  edm::EDGetTokenT<std::vector<PCaloHit>> simHitsTokenHEback_;
+
+  HGCalClusteringDummyImpl dummyClustering_;
+  HGCalShowerShape showerShape_;
+
+  HGCalTriggerTools triggerTools_;
+};
+
+CaloTruthCellsProducer::CaloTruthCellsProducer(edm::ParameterSet const& config)
+    : makeCellsCollection_(config.getParameter<bool>("makeCellsCollection")),
+      caloParticlesToken_(consumes<CaloParticleCollection>(config.getParameter<edm::InputTag>("caloParticles"))),
+      triggerCellsToken_(
+          consumes<l1t::HGCalTriggerCellBxCollection>(config.getParameter<edm::InputTag>("triggerCells"))),
+      simHitsTokenEE_(consumes<std::vector<PCaloHit>>(config.getParameter<edm::InputTag>("simHitsEE"))),
+      simHitsTokenHEfront_(consumes<std::vector<PCaloHit>>(config.getParameter<edm::InputTag>("simHitsHEfront"))),
+      simHitsTokenHEback_(consumes<std::vector<PCaloHit>>(config.getParameter<edm::InputTag>("simHitsHEback"))),
+      dummyClustering_(config.getParameterSet("dummyClustering")) {
+  produces<CaloToCellsMap>();
+  produces<l1t::HGCalClusterBxCollection>();
+  produces<l1t::HGCalMulticlusterBxCollection>();
+  if (makeCellsCollection_)
+    produces<l1t::HGCalTriggerCellBxCollection>();
+}
+
+CaloTruthCellsProducer::~CaloTruthCellsProducer() {}
+
+void CaloTruthCellsProducer::produce(edm::Event& event, edm::EventSetup const& setup) {
+  edm::Handle<CaloParticleCollection> caloParticlesHandle;
+  event.getByToken(caloParticlesToken_, caloParticlesHandle);
+  auto const& caloParticles(*caloParticlesHandle);
+
+  edm::Handle<l1t::HGCalTriggerCellBxCollection> triggerCellsHandle;
+  event.getByToken(triggerCellsToken_, triggerCellsHandle);
+  auto const& triggerCells(*triggerCellsHandle);
+
+  dummyClustering_.eventSetup(setup);
+  showerShape_.eventSetup(setup);
+  triggerTools_.eventSetup(setup);
+
+  edm::ESHandle<HGCalTriggerGeometryBase> geometryHandle;
+  setup.get<CaloGeometryRecord>().get(geometryHandle);
+  auto const& geometry(*geometryHandle);
+
+  std::unordered_map<uint32_t, CaloParticleRef> tcToCalo;
+
+  std::unordered_map<uint32_t, double> hitToEnergy(makeHitMap(event, setup, geometry));  // cellId -> sim energy
+  std::unordered_map<uint32_t, std::pair<double, double>>
+      tcToEnergies;  // tcId -> {total sim energy, fractioned sim energy}
+
+  for (auto const& he : hitToEnergy) {
+    DetId hitId(he.first);
+    uint32_t tcId;
+    try {
```

Comment:
Thanks @kpedro88. This plugin is not supposed to run in production, but we can change that. @yiiyama could you have a look?

Reply:
@jbsauvan try/catch is here because the various HGCal trigger geometry subclasses throw in case a cell that's not mapped to a trigger cell is found. I guess in principle we shouldn't have any such non-mapped cells? But I did encounter a few cases and that's why the try/catch was introduced, because seeing such cells is not a fatal problem for this plugin (can just skip over them). The options here would be

1. Try to retain this "skip over" behavior by
   - add an option to the trigger geometry subclasses to fail silently (return trigger cell id 0 for example)
   - essentially copy-paste various geometry subclass code here, and take out the throwing part
1. Let the plugin crash whenever there is an unmapped cell (i.e. just bring the `getTriggerCellFromCell` line out of the try block and don't catch anything). We'll then identify the offending cells and do something about them.
1. Because the plugin will not be run in production, leave the code as is

I'd vote for 2. or 3. What do you think?

Reply:
@yiiyama 
Ok, I was not aware of this issue. If there are problems of mapping it should not continue running. So I would remove the `try/catch` block (option 2.)

Reply:
Thanks for addressing this - my basic view on try/catch is that if some upstream code is throwing an exception, it's probably doing that for a reason, so catching it and downgrading the severity may lead to undesirable results. If we think the upstream code shouldn't throw, that code itself should be changed. (Not in this case, as indicated.)

Reply:
@jbsauvan @yiiyama in general try - catch is against our code rules, and indeed it is marked as an issue in the code checks. If there is a simple protection to the code it would be preferable, I agree with @kpedro88 that try - catch is just "hiding a problem under the carpet"

Reply:
@fabiocos thank you for checking in. We removed the try/catch already in ebb546f (merged below).